### ISSUE TITLE
Crossfade operator messes up source.elapsed(s) and source.duration(s)

### ISSUE THREAD

**Issue by gAlleb** [savonet/liquidsoap#3404]
**Describe the bug**

If you have sources before crossfade operator and want to use source after applying crossfade operator as a source of time data then time data of tracks will be inaccurate — in an increasing pattern:
`source.elapsed(s) and source.duration(s) ` will always increase
`source.remaining(s)` will not be affected

**To Reproduce**
```
radio = switch([({ 7h-22h }, dayshift), ({ 22h-7h }, nightshift)])
radio = crossfade(smart = false, fade_out=3.00, fade_in=0.00, duration=3.00)
```
```
def print_source_info(s)
    elapsed = format_time(source.elapsed(s))
    remaining = format_time(source.remaining(s))
    duration = format_time(source.duration(s))
    id = source.id(s)
    print("id=#{id} elapsed=#{elapsed} remaining=#{remaining} duration=#{duration}")
end
thread.run({print_source_info(radio)}, every=5.)
```
```
id=switch.23 elapsed=01m:29s remaining=00m:26s duration=01m:56s
id=switch.23 elapsed=01m:31s remaining=00m:24s duration=01m:56s
id=switch.23 elapsed=01m:33s remaining=00m:22s duration=01m:56s
id=switch.23 elapsed=01m:35s remaining=00m:20s duration=01m:56s
id=switch.23 elapsed=01m:37s remaining=00m:18s duration=01m:56s
id=switch.23 elapsed=01m:39s remaining=00m:16s duration=01m:56s
id=switch.23 elapsed=01m:41s remaining=00m:14s duration=01m:56s
id=switch.23 elapsed=01m:43s remaining=00m:12s duration=01m:56s
id=switch.23 elapsed=01m:45s remaining=00m:10s duration=01m:56s
id=switch.23 elapsed=01m:47s remaining=00m:08s duration=01m:56s
id=switch.23 elapsed=01m:49s remaining=00m:06s duration=01m:56s
id=switch.23 elapsed=01m:51s remaining=00m:04s duration=01m:56s
id=switch.23 elapsed=01m:53s remaining=00m:02s duration=01m:56s
id=switch.23 elapsed=01m:55s remaining=00m:00s duration=01m:56s
Track changed
id=switch.23 elapsed=01m:57s remaining=00m:01s duration=01m:59s
id=switch.23 elapsed=01m:59s remaining=04m:19s duration=06m:18s
id=switch.23 elapsed=02m:01s remaining=04m:17s duration=06m:18s
id=switch.23 elapsed=02m:03s remaining=04m:15s duration=06m:18s
id=switch.23 elapsed=02m:05s remaining=04m:13s duration=06m:18s
id=switch.23 elapsed=02m:07s remaining=04m:11s duration=06m:18s
```

**Expected behavior**
Crossfade operator  should not affect source.elapsed(s) and source.duration(s) time data

As for now to fetch correct time data one must use source before applying `crossfade operator`

**Version details**
- OS: Ubuntu 20.04
- Liquidsoap 2.2.1

**Install method**
opam

Thanks to @vitoyucepi for pointing it out!

**Comment by Moonbase59** [comment]
We’ve been having that for some time (and been trying to find ways for progress indicators, hee hee). One could argue that the new source after crossfading/mixing is essentially one new, long track. Apparently this is what LS does, so the previous single track info gets lost.

**Comment by toots** [comment]
Yes I can concur.  How would y'all think we should mark a single track change in the presence of a crossfade? What would be the most expected behavior? 

For reference:

![Drawing](https://github.com/savonet/liquidsoap/assets/871060/07df6bf7-db0f-4821-a3b5-075e5aa7e69f)

**Comment by pozhiloy-enotik** [comment]
@toots I might insist that this is a somewhat incorrect example. An example of a professional crossfade is as follows:
![image](https://github.com/savonet/liquidsoap/assets/49619526/b6ff8224-b4db-48a4-8580-a8a42028f37e)

**Comment by toots** [comment]
@pozhiloy-enotik I can see that. I think that the problem is that we also rely on events triggered by track start as well as track end and, as of now, we only have one type of even, `track_mark` so, the above, if we were to mark the end of track there, we would also implicitly say that the new track starts there are well, which would mess up events related to the new track's start.

I'll see if we could do better.

**Comment by toots** [comment]
Actually, to give another example: what would you return for `source.elapsed` during this part of your timeline:

![269422931-b6ff8224-b4db-48a4-8580-a8a42028f37e](https://github.com/savonet/liquidsoap/assets/871060/f88b78c2-addf-4f33-a252-e724ff64dfa9)

**Comment by Moonbase59** [comment]
This is actually complicated, since we don’t really have _track-based_ counts.

Personally, I’d probably count out "source.elapsed" until "End of track" above, but introduce a "help counter" that starts with track "blue" starting. After track "red" is finished, I’d update "source.elapsed" to the value of the "help counter" and count on, so as to show the real playtime of the current track.

This is of course only a workaround, but might mirror "reality" a bit more closely.
